### PR TITLE
Easier installation for GPU/TPU

### PR DESCRIPTION
**Pull request by lumip**
Currently, numpyro installation instructions for GPU support ask the user to "first install `jax` and `jaxlib` with CUDA support." This is inconvenient and prone to lead to errors as the user then needs to watch out to install compatible versions of `jax` and `jaxlib`. Similar for TPU support.

Lately, JAX has improved its installation scripts by adding optional targets for GPU/TPU support that automatically install a CUDA/TPU version of `jaxlib` that is compatible with the installed version of `jax`, starting from version `0.2.11`. This PR adds the same targets to numpyro (and internally simply relies on the identical jax targets).

Installing numpyro for CUDA is now possible with
```
pip install numpyro.[cuda101] -f https://storage.googleapis.com/jax-releases/jax_releases.html
```
assuming CUDA 10.1 (similar for 10.2, 11.1, ..)
and for TPU
```
pip install numpyro.[tpu] -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
```

The additional package index URLs are currently required by JAX and are a bit unfortunate, but overall this is still simpler and less error prone than the current approach.

### CODE REVIEW

**Comment by fehiepsi** [comment]
Hi @lumip, this is great! I just wonder if https://github.com/google/jax/pull/7146 is relevant before merging. Currently, I don't have a GPU system to test this yet. :(

**Comment by lumip** [comment]
Hey @fehiepsi : I opened the linked PR in an attempt to make this even nicer by being able to drop the additional package index url given to pip. However, due to the PyPI policy of not allowing packages to install dependencies that are not listed in a package index, the changes made by that PR were unfortunately not possible. That is, unless the CUDA versions of jaxlib become available on PyPI, this is currently the best we can do..

File: setup.py
Comment (fehiepsi) [inline setup.py:18]: nit: no space after `">=`

Does the latest jax release support `jax[cpu]` syntax? If not, I think we can wait until after the next numpyro release.

File: README.md
Comment (fehiepsi) [inline README.md:200]: Maybe we can rephrase this to `... setup the TPU backend as detailed here and use pip to install NumPyro as above`?

File: setup.py
Comment (fehiepsi) [inline setup.py:41]: Do you know what happens if `jax`, `jaxlib` GPU version is available and we want to update NumPyro
```
pip install -U numpyro
```
Will CPU versions replace the available GPU versions?

**Comment by fehiepsi** [comment]
Understood! Thanks for explaining, @lumip!

File: setup.py
Comment (lumip) [reply setup.py:18]: The `[cpu]` and `[tpu]` targets for jax are consistently available since jax version 0.2.11. However, I just noted that prior to 0.2.13 the GPU targets are `[cuda-xxx]` and since 0.2.13 its `[cudaxxx]`, which is a bit annoying. It might make sense to wait with this until we the jax dependency is bumped to 0.2.13 or above.

File: setup.py
Comment (lumip) [reply setup.py:41]: This is indeed what happens (provided there was a change in the jax version NumPyro depends on). Updates would need to also specify the cuda option (and the jax release repo URL) in order not to replace with the CPU version.
One option would be to not install jaxlib unless the user specifies the target platform (which apparently what the folks at jax are going for). This means that jaxlib does not get replaced unintentionally but carries the risk of ending up with updated versions of numpyro and jax and an outdated (incompatible) jaxlib from the initial installation..

File: README.md
Comment (lumip) [reply README.md]: I have rephrased this now

**Comment by lumip** [comment]
One further thing to consider with installation in general is that jaxlib does not maintain backward compatibility, so only constraining the version from below (as done previously in numpyro and would be the case through the jax[whatever] installation from this PR) means that users installing older versions will run into issues with high probability. I have raised an [issue](https://github.com/google/jax/issues/7174) with the folks at JAX about this, let's see what they think.

File: setup.py
Comment (fehiepsi) [reply setup.py:41]: So I think we can remove this `cpu` option and set
```
f"jax>={_jax_version_constraints}",
"jaxlib>=0.1.62",
```
as previously. Since the next release, we will relax `jax/jaxlib` version - when people use numpyro in colab, they won't need to worry about jax/jaxlib installation (colab environment already has those versions)

File: README.md
Comment (fehiepsi) [reply README.md]: Thanks!

File: setup.py
Comment (fehiepsi) [reply setup.py]: Please feel free to pump the minimum jax version to `0.2.13` and jaxlib to `0.1.65`. Currently, colab is using jax 0.2.13 and jaxlib 0.1.66 so its users won't have to reinstall (which potentially leads to conflicts) jax/jaxlib there.

**Comment by fehiepsi** [comment]
> users installing older versions will run into issues with high probability

I think that if the new versions are available, we can just inform them to update the packages (if we suspect that the issue is due to an old release of jaxlib).

File: README.md
Comment (fehiepsi) [inline README.md:204]: I think this command can be confusing. If I understand correctly, those people who want to run `pip install numpyro[tpu] ...` are using Cloud TPU VM (not Colab) and they want to skip [this step](https://cloud.google.com/tpu/docs/jax-quickstart-tpu-vm) in the Cloud TPU VM JAX setup process. I'm not sure we should encourage such simplification. In my opinion, we should encourage users to follow the upstream quickstart, test the environment, then install numpyro with `pip install numpyro`.

**Comment by lumip** [comment]
> I think that if the new versions are available, we can just inform them to update the packages (if we suspect that the issue is due to an old release of jaxlib).

The problem is exactly the other way around: An installation of, say, numpyro v0.6.0, in the future may pull any newer jaxlib (or even jax) version which may well not be compatible, as neither make any guarantees about backward compatibility. I think it would be prudent to make sure that if the user wants to install a specific version of numpyro, they get it with working dependencies instead of being told to upgrade.
This is now fixed in jax for installation through the platform targets (`cpu`, `cuda`, `tpu`) by pinning a known compatible jaxlib version. However, numpyro installation might still pull an incompatible jax version (because it only constraints that from below).

File: README.md
Comment (lumip) [reply README.md:204]: Alright, I'll revisit this.

**Comment by fehiepsi** [comment]
I see your point. In the past, we pinned jax/jaxlib versions to avoid breaking changes. But we think that jax/jaxlib has been more stable and requiring users to install `master` branch of numpyro is a bit inconvenient now. I'm not sure what jaxlib does after all so I can't tell how much its new releases will affect numpyro users. How about having `numpyro[cpu]` option to install a jaxlib version that is "assumed-by-jax-devs" compatible with the latest jax release? For the default installation, I think it is best, for now, to only check for jax/jaxlib requirements (and, as a pip policy, pip will try to install the latest versions if these dependencies are not available). Users can choose to `pip install numpyro[cpu]` to pin jaxlib if they want. What do you think?

File: README.md
Comment (lumip) [reply README.md]: Alright, maybe this is fine now?

**Review by fehiepsi — COMMENTED** [review]
Yeah, let's see how things go. We might skip edge cases for now (e.g. some breaking changes that happen with any software or someone wants to upgrade jaxlib but keep the old jax version,...) CI can help us to detect the issues early so personally, I don't worry too much about them. Regarding `cpu`, I don't have a better name. It is fine I guess (at least it is concise :D).

File: README.md
Comment (lumip) [reply README.md]: fixed

File: README.md
Comment (lumip) [reply README.md]: done

File: README.md
Comment (lumip) [reply README.md]: done

**Review by fehiepsi — APPROVED** [review]
This is extremely helpful, thanks @lumip! Please let me know if you have any concerns that need further discussion or this is ready to merge. :)

**Comment by lumip** [comment]
I think I'm happy with it now. I can squash it into a single commit still..

**Comment by lumip** [comment]
Thanks for the review and the discussion to make improve this!

**Pull request merged**
